### ISSUE TITLE
readObject with type = wrapped throws for a commit object

### ISSUE THREAD

**Issue by chmac** [isomorphic-git/isomorphic-git#1240]
Example reproduction:

```typescript
        const wrappedObject = await git.readObject({
          fs,
          oid: 'objectId of a commit',
          format: 'wrapped',
        });
```

When I try the above, I get the following exception:

```
TypeError: Cannot set property 'oid' of undefined
    at Object.readObject (node_modules/isomorphic-git/index.cjs:11827:16) {
  caller: 'git.readObject'
```

Digging into this a bit, I find that here `result` is undefined:

https://github.com/isomorphic-git/isomorphic-git/blob/f2e38053733f4c8ca4b7784d5c2a56addcf36bf4/src/api/readObject.js#L228-L235

It seems like `_readObject()` is returning `undefined`. Maybe there's a bug in this switch statement?

https://github.com/isomorphic-git/isomorphic-git/blob/071e8ec92f8f0e07a8a067082f0a0f316f3ba026/src/storage/readObject.js#L61-L70

Going into this switch statement, `format = 'wrapped'` while `result.format = 'content'`. So we hit the `case 'content':`, but then not the nested `if`, which means we hit `break`, and so we skip the `default` branch, but there's no closing `return` in the function, and so `result` is never returned, and we get back `undefined` instead.

I'm not sure what's going on here. Perhaps for a commit object, `wrapped` and `content` are synonymous because there is no "wrapped" version of a commit?

**Comment by jcubic** [comment]
Hey, thanks for reporting a bug and great production and find the root cause. Maybe you can create a PR with a fix.

Sorry for the late reply, but I've become a maintainer recently and and I've started to handling issues and PR. The original author doesn't have much time to work on the project so I've started helping, but since I also don't have much time I don't know if I will be able to write code. So if you want to help and become contributor to this Open Source project, your help is welcome.

**Comment by balajianbalagan** [comment]
Hey, can I work on this issue? I'm new to Open source contribution but good at JS and programming in general

**Comment by jcubic** [comment]
Go ahead, if you can fix the issue that would be great. But I'm not sure what would be the right code to fix the issue. You will need to figure it out on own, which is the most difficult part of the issue.

**Comment by dead-end** [comment]
@balajianbalagan I think the code changed since the issue was reported. I get the mentioned error:

```
TypeError: Cannot set properties of undefined (setting 'oid')
```
when I pass a wrong format:
```
        const wrappedObject = await git.readObject({
          fs,
          oid: 'objectId of a commit',
          format: 'joho',
        });
```
In the original code I think the issue occurred, when you pass a format `content` on a wrapped objects and vice versa. If you want to contribute code, add the parsing of the format parameter, add a unit test for the parsing to the `test-readObject.js` And then you can add two unit tests for the cases:
- parameter `content` object `wrapped`
- parameter `wrapped` object `content`

**Comment by balajianbalagan** [comment]
Hey, @jcubic! yes so I looked at the code and I need to know which approach we can take to solve this as @dead-end mentioned the error occurs only if passed the wrong format, so I thought of adding a parameter check in the readObject.js file for the format and added an error object called InvalidFormatError.js, so do we want to throw an error if an invalid format passed (that's what I'm doing right now) or We could take a default format value and inform the client that "the format is not recognized reading using the 'default - parsed' method"?
I have tried reading objects of different types it is working 
![image](https://github.com/isomorphic-git/isomorphic-git/assets/42893328/68104614-1e2b-495a-ab84-6479bf6c1d23)

**Comment by jcubic** [comment]
I'm not sure if I understand but I would just check if the format is valid and if not throw a proper exception:
```javascript
if ([...alowedFormats].includes(_format)) {
   throw new SomeError(...);
}
```

It will probably require creating a new type of exception, similar to others that are already in the code.

**Comment by balajianbalagan** [comment]
Yes that is what im intending to do! I'll implement this.

**Comment by limond** [comment]
I think the original post and the discussion do not really fit.

It should be perfectly fine to ask for a `wrapped` commit object through the `readObject` function. There is even a [test](https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/__tests__/test-readObject.js#L99) for this.

The discussion then went on to define allowed formats but I also can reproduce `TypeError: Cannot set property 'oid' of undefined` with a valid format `wrapped` on commit objects. The interesting part is that this only happens if a commit is packed and I believe this was the original cause here.

The documentation states the following for the returned object:

> format is usually, but not always, the format you requested. Packfiles do not store each object individually compressed so if you end up reading the object from a packfile it will be returned in format 'content' even if you requested 'deflated' or 'wrapped'.

For `deflated` this works correctly. For `wrapped` this does not work.

# Analysis of the code of `_readObject`

**TLDR**: `_readObject` tries to find a loose object. As a fallback it tries to find a packed object. Loose objects always have the format `deflated` (see `readObjectLoose`), packed objects have the type `content` (see `readObjectPacked`) as *individual* objects are not compressed. If a user has requested a `deflated` object, the result (either loose and `deflated` or packed and `content`) is returned early: https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L54-L56
 If a user has requested `wrapped` for a packed object, all outer `if`s that check the `result.format` will not be hit. The first `if` that is hit is this one: https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L79-L82

The `result.format` is `content` (as with all packed objects) but since `content` was not requested but `wrapped` this returns `undefined` and leads to a TypeError in the command above which is a bug as the documentation states `wrapped` packed objects will be returned as `content`.

**Deeper analysis**:

The following table shows what happens to an object that is either loose or packed depending on which format was requested:

<table>
  <tr>
    <th>Lines</th>
    <th colspan="3">loose object</th>
    <th colspan="3">packed object</th>
  </tr>
  <tr>
    <td></td>
    <td>format = deflated</td>
    <td>format = wrapped</td>
    <td>format = content</td>
    <td>format = deflated</td>
    <td>format = wrapped</td>
    <td>format = content</td>
  </tr>
  <tr>
    <td><a href="https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L54-L56">L54-L56</></td>
    <td>object is returned with result.format = deflated</td>
    <td>-</td>
    <td>-</td>
    <td>object is returned with result.format = content as per documentation of the corresponding command</td>
    <td>-</td>
    <td>-</td>
  </tr>
  <tr>
    <td><a href="https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L58-L61">L58-L61</></td>
    <td></td>
    <td>Object is inflated</td>
    <td>Object is inflated</td>
    <td></td>
    <td>-</td>
    <td>-</td>
  </tr>
  <tr>
    <td><a href="https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L63-L77">L63-L77</></td>
    <td></td>
    <td>object is returned with result.format = wrapped</td>
    <td>object is checked and unwrapped</td>
    <td></td>
    <td>-</td>
    <td>-</td>
  </tr>
  <tr>
    <td><a href="https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L79-L82">L79-L82</></td>
    <td></td>
    <td></td>
    <td>object is returned with result.format = content</td>
    <td></td>
    <td>🚨 undefined is returned (and causes an error in the command above)</td>
    <td>object is returned with result.format = content</td>
  </tr>
</table>

# Opinion

There are a few things I do not understand:
- Why is the result of a packed object not returned immediately? There is no reason to let this function continue with a packed object. For all allowed formats `content`, `wrapped` and `deflated` the result should be of format `content` as per documentation. We could return the result as soon as `readObjectPacked` has returned an object of format `content`.
- The structure of the `if`s is confusing for me. Two outer `if`s
https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L63 https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L79 are always true for loose objects. Inside each `if` the `result.format` is always set so that the next `if` will be true for loose objects. If we would return early for packed objects, this could be simplified a lot. Also the second term https://github.com/isomorphic-git/isomorphic-git/blob/bbcdda7d0cd401c5a79d1f372ea31f968e44e93f/src/storage/readObject.js#L63-L64 here (`result.format === 'wrapped'`) is superfluous and adds more confusion.

# Suggested changes

I will make a PR to get feedback for the suggested changes.

**Comment by luantaraschi** [comment]
I went to fix this and could not reproduce the reported symptom on `main` (`ab6ff20`). Posting what I measured, in case it helps someone decide whether to close it.

`format: 'wrapped'` on a commit works. There is a passing test for exactly that case in `__tests__/test-readObject.js`, and it returns `format: 'wrapped'` as expected.

@dead-end's variant, passing a format that is not valid, no longer produces `TypeError: Cannot set properties of undefined`. It throws an `InternalError` now.

What I did find while checking is a difference the caller cannot predict. I ran the same five calls against the `test-readObject` fixture:

| oid | format | result |
|---|---|---|
| loose commit | `wrapped` | returns `format: 'wrapped'` |
| loose commit | `joho` | throws `InternalError` |
| packed commit | `joho` | **returns `format: 'content'`, no error** |
| packed commit | `wrapped` | returns `format: 'content'` |
| packed commit | `deflated` | returns `format: 'content'` |

The last two are intentional and documented (`// packed objects will always be returned as 'content' format`, asserted in `test-readObject.js`). The third one looks like a side effect of that rule rather than a decision: `_readObject` returns the packed result before it ever looks at `format`, in `src/storage/readObject.js`:

```js
if (!result) {
  result = await readObjectPacked({ fs, cache, gitdir, oid, getExternalRefDelta })
  if (!result) throw new NotFoundError(oid)
  // Directly return packed result, as specified: packed objects always return the 'content' format.
  return result
}
```

The validation that catches a bad format sits at the bottom of the function, on the loose path only:

```js
throw new InternalError(`invalid requested format "${format}"`)
```

So `readObject({ oid, format: 'typo' })` throws or succeeds depending on whether that object happens to be packed, which is a function of when the repository was last repacked and not something the caller controls.

If you think that is worth fixing, validating `format` at the top of `_readObject` would make it behave the same either way, and would not change anything about the packed-returns-content rule. Happy to send that as a small PR if you want it. Otherwise this issue looks stale to me and could probably be closed.

**Comment by jcubic** [comment]
@luantaraschi You can create a PR to fix this if you want.
